### DISCUSSION TITLE
Handling of system release version with other sources (e.g. subscription management)

### DISCUSSION THREAD

**Discussion by ptoscano** [rpm-software-management/dnf5#423] [General]
`dnf` currently manages the version of the system (represented by the `$releasever` variable) on its own, using the version of the `*-release` package (e.g. `redhat-release`, `centos-release`, etc) as default, with the possibility to override it via variables.

`subscription-manager` is the tool currently in use in RHEL to let users registers their systems. One of the functions it has is managing the version of the system from the "subscription" point of view. This is functionality is exposed to the user as separate subcommand:
```
# subscription-manager release --help
usage: subscription-manager release [OPTIONS]

Configure which operating system release to use

optional arguments:
  -h, --help            show this help message and exit
  --proxy PROXY_URL     proxy URL in the form of hostname:port
  --proxyuser PROXY_USER
                        user for HTTP proxy with basic authentication
  --proxypassword PROXY_PASSWORD
                        password for HTTP proxy with basic authentication
  --noproxy NO_PROXY    host suffixes that should bypass HTTP proxy
  --no-progress-messages
                        Do not display progress messages
  --show                shows current release setting; default command
  --list                list available releases
  --set RELEASE         set the release for this system
  --unset               unset the release for this system
```
Determining the list of available releases for a certain system requires at least contacting the registration server (there is more to it, I'm keeping this simple in this context), and so also setting/unsetting the release (which is saved on the registration server).

While this approach "works", in practice it has a lot of drawbacks:
- the concept of "release" that dnf and `subscription-manager` have are totally separate and alien to each other
- `subscription-manager` is basically managing the base version of the OS... which is what dnf does already managing the installed software
- when the release is changed via `subscription-manager release --set/--unset`, `subscription-manager` rewrites its own `redhat.repo` file replacing the value of `$releasever` (the issues with the repository management of `subscription-manager` are a separate discussion, i.e. #422); this means that other repositories are not following the release set for the system

Hence, it would be nice if dnf had a clean way for external sources to help determine the release of the system using the plugin infrastructure. The request here is not to implement the logic of `subscription-manager` in dnf, of course, but rather something along the lines of:
- a plugin would be able to provide a source of the release of the system
- dnf would either use the source of that plugin as default, e.g. when configured to do so
- changing the the release to use (e.g. via `--releasever`) might be constrained to the releases available
- possibly a new subcommand in dnf to manage the release cleanly? this also would need to notify the plugin about it, so the plugin could be able to accept/reject the release change and do additional work

**Comment by evan-goode** [comment]
> a plugin would be able to provide a source of the release of the system
> dnf would either use the source of that plugin as default, e.g. when configured to do so
> changing the the release to use (e.g. via --releasever) might be constrained to the releases available
> possibly a new subcommand in dnf to manage the release cleanly? this also would need to notify the plugin about it, so the plugin could be able to accept/reject the release change and do additional work

I'm skeptical that we need to add more releasever-specific functionality to DNF. I think a lot of this might be achieved using DNF5's existing functionality for plugin hooks, and vars.

Spitballing, what if:

- `redhat.repo` uses the `$releasever` variable, so that it doesn't need to be regenerated when the releasever changes.
- User can list available releasevers for their system using `subscription-manager release --list`
- User can set the releasever for a one-off command (e.g. upgrade/system-upgrade) with `dnf5 upgrade --releasever 11.2`
- If desired, a user could override the releasever permanently with `dnf5 config-manager setvar releasever=11.2`. (I'm not sure what the use case for this is, I'd assume overriding the releasever is mostly useful for upgrading to a new version, after which /etc/os-release would have the new releasever)
- Some libdnf5 subscription-manager plugin hooks into dnf5's repository configuration step, detects when a releasever is set that differs from the current subscription configuration, and does the appropriate registration/telemetry. The plugin would also cause DNF5 to error if an unavailable releasever was selected.

That seems like a reasonable architecture to me, but I'm not really familiar with the reasons subscription-manager is the way it is.
